### PR TITLE
Refactor compiler: only use AppElement for ViewContaines

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

**Comment by tbosch** [comment]
Only the last 4 commits. The ones before are from https://github.com/angular/angular/pull/12635

File: modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export function bindDirectiveWrapperLifecycleCallbacks(
       DirectiveWrapperExpressions.ngOnDestroy(dir.directive, directiveWrapperIntance));
   compileElement.view.detachMethod.addStmts(DirectiveWrapperExpressions.ngOnDetach(
       dir.hostProperties, directiveWrapperIntance, o.THIS_EXPR,
-      compileElement.component ? compileElement.appElement.prop('componentView') : o.THIS_EXPR,
+      compileElement.compViewExpr ? compileElement.compViewExpr : o.THIS_EXPR,
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts]: Use `||`

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 2]
```diff
@@ -110,7 +110,7 @@ export function bindDirectiveHostProps(
   compileElement.view.detectChangesRenderPropertiesMethod.addStmts(
       DirectiveWrapperExpressions.checkHost(
           directiveAst.hostProperties, directiveWrapperInstance, o.THIS_EXPR,
-          compileElement.component ? compileElement.appElement.prop('componentView') : o.THIS_EXPR,
+          compileElement.compViewExpr ? compileElement.compViewExpr : o.THIS_EXPR,
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Use `||`

File: modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts]: Done

File: modules/@angular/compiler/src/view_compiler/property_binder.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Done

**Comment by code-tree** [comment]
I'm getting

`Error: Error at share.service.ngfactory.ts:9:26: Cannot find module '@angular/core/src/linker/element'.`

I'm assuming it is because of this change. Anything I need to change in my app to account for this?

**Comment by tbosch** [comment]
You are probably mixing different versions for @angular/compiler and
@angular/core...

**Comment by code-tree** [comment]
Thanks @tbosch. I thought I already had updated them all, but tried upgrading again and now it works. cheers

**Pull request merged**